### PR TITLE
Preserve case of the last name when creating or editing a user

### PR DESCRIPTION
**Pull request by jlnat**
Whenever creating an new user, or editing an existing one, the casing on the last name will be preserved.
This does not have any impact on displaying the name everywhere else on the application, as that will still be displaying as `LASTNAME, Firstname`

Closes [AB#1193](https://dev.azure.com/ncia-anet/2aa083a5-af3d-44e1-8c7b-6e9e6b124d91/_workitems/edit/1193)

#### User changes
- Ability to edit the last name of a user without any constraints regarding the casing

#### Superuser changes
- None

#### Admin changes
- None

#### System admin changes
- [ ] application.yml or anet-dictionary.yml needs change
- [ ] db needs migration
- [ ] documentation has changed
- [ ] graphql schema has changed

### Checklist
- [x] described the user behavior in PR body
- [x] referenced/updated all related issues
- [x] commits follow a `repo#issue: Title` title format and [these 7 rules](https://chris.beams.io/posts/git-commit/)
- [x] commits have a [clean history](https://epage.github.io/dev/commits/), otherwise PR may be squash-merged
- [x] added and/or updated unit tests
- [x] added and/or updated e2e tests
- [ ] added and/or updated data migrations
- [x] updated documentation
- [x] resolved all build errors and warnings
- [ ] opened debt issues for anything not resolved here

### CODE REVIEW

File: client/src/pages/admin/merge/MergePeople.tsx [hunk 1]
```diff
@@ -497,12 +497,19 @@ const PersonColumn = ({
...
+          }
           overlayColumns={["name"]}
           overlayRenderRow={PersonSimpleOverlayRow}
           filterDefs={peopleFilters}
           onChange={value => {
-            value?.fixupFields()
+            if (value) {
+              value.name = Person.militaryName(value.name)
+              value.fixupFields()
+            }
```
Comment (gjvoosten) [inline client/src/pages/admin/merge/MergePeople.tsx:512]: Unnecessary change, please revert.

File: client/src/models/Person.ts [hunk 2]
```diff
@@ -360,6 +360,14 @@ export default class Person extends Model {
     return r
   }
 
+  static militaryName(name) {
+    const parsedName = Person.parseFullName(name)
```
Comment (gjvoosten) [inline client/src/models/Person.ts:364]: Be aware that `name` can be null here, so adapt `parseFullName` below with:
```javascript
@@ -369,8 +369,8 @@ export default class Person extends Model {
   }
 
   static parseFullName(name) {
-    const delimiter = name.indexOf(Person.nameDelimiter)
-    let lastName = name
+    const delimiter = name?.indexOf(Person.nameDelimiter)
+    let lastName = name ?? ""
     let firstName = ""
 
     if (delimiter > -1) {
```

**Review by gjvoosten — CHANGES_REQUESTED** [review]

File: client/src/pages/admin/merge/MergePeople.tsx [hunk 3]
```diff
@@ -497,12 +497,19 @@ const PersonColumn = ({
         <AdvancedSingleSelect
           fieldName="person"
           placeholder="Select a person to merge"
-          value={person}
+          value={
+            person
+              ? { ...person, name: Person.militaryName(person.name) }
+              : person
+          }
```
Comment (gjvoosten) [inline client/src/pages/admin/merge/MergePeople.tsx:504]: Unnecessary change, please revert.

File: client/src/pages/people/Edit.tsx [hunk 4]
```diff
@@ -117,9 +117,10 @@ const PersonEdit = ({ pageDispatchers }: PersonEditProps) => {
     }
   }
   const person = new Person(data ? data.person : {})
+  const personName = Person.militaryName(person.name)
```
Comment (gjvoosten) [inline client/src/pages/people/Edit.tsx:120]: This variable can be in-lined.

File: client/src/pages/admin/merge/MergePeople.tsx (hunk 1 above)
Comment (jlnat) [reply client/src/pages/admin/merge/MergePeople.tsx:512]: I believe we want to maintain the previous behaviour where the last name is capitalized, right?

File: client/src/pages/admin/merge/MergePeople.tsx (hunk 3 above)
Comment (jlnat) [reply client/src/pages/admin/merge/MergePeople.tsx:504]: If we want to maintain the previous behaviour where the last name is capitalized on the Merge People page, then I believe we do need this.
This line in specific covers edge cases when we come to this page with a name selected already

File: client/src/pages/people/Edit.tsx (hunk 4 above)
Comment (jlnat) [reply client/src/pages/people/Edit.tsx:120]: It can but I would go over the line limit and need to make a new line regardless, so I thought this was cleaner.
Do you still want me to remove the const?

Comment (gjvoosten) [reply client/src/pages/people/Edit.tsx:120]: Yes please; it localises the change to just the line that is affected.

File: client/src/pages/admin/merge/MergePeople.tsx (hunk 1 above)
Comment (gjvoosten) [reply client/src/pages/admin/merge/MergePeople.tsx:512]: If we really want to do that, something like:
```javascript
           objectType={Person}
           valueKey="name"
+          valueFunc={(v, k) => k === "name" ? Person.militaryName(v?.[k]) : v?.[k]}
           fields={Person.allFieldsQuery}
           addon={PEOPLE_ICON}
```
would be a better solution, but we would have to apply this to all AdvancedSingleSelect's for person.

File: client/src/pages/admin/merge/MergePeople.tsx (hunk 3 above)
Comment (gjvoosten) [reply client/src/pages/admin/merge/MergePeople.tsx:504]: See https://github.com/NCI-Agency/anet/pull/5006#discussion_r1871461728

File: client/src/pages/admin/merge/MergePeople.tsx [hunk 5]
```diff
@@ -497,16 +497,24 @@ const PersonColumn = ({
...
           overlayColumns={["name"]}
           overlayRenderRow={PersonSimpleOverlayRow}
           filterDefs={peopleFilters}
           onChange={value => {
-            value?.fixupFields()
+            if (value) {
+              value.fixupFields()
+            }
```
Comment (gjvoosten) [inline client/src/pages/admin/merge/MergePeople.tsx:511]: Unnecessary change, please revert.

File: client/src/pages/admin/merge/MergePeople.tsx [hunk 6]
```diff
@@ -497,16 +497,24 @@ const PersonColumn = ({
         <AdvancedSingleSelect
           fieldName="person"
           placeholder="Select a person to merge"
-          value={person}
+          value={
+            person
+              ? { ...person, name: Person.militaryName(person.name) }
+              : person
+          }
```
Comment (gjvoosten) [inline client/src/pages/admin/merge/MergePeople.tsx:504]: Unnecessary change, please revert.

File: client/src/pages/admin/merge/MergePeople.tsx [hunk 7]
```diff
@@ -497,16 +497,24 @@ const PersonColumn = ({
...
-            value?.fixupFields()
+            if (value) {
+              value.fixupFields()
+            }
             dispatchMergeActions(setMergeable(value, align))
           }}
           objectType={Person}
           valueKey="name"
+          valueFunc={(v, k) =>
+            k === "name" ? Person.militaryName(v?.[k]) : v?.[k]}
```
Comment (gjvoosten) [inline client/src/pages/admin/merge/MergePeople.tsx:517]: Unnecessary change, please revert.

**Review by gjvoosten — CHANGES_REQUESTED** [review]

**Review by gjvoosten — APPROVED** [review]
